android/modules/gesture/src/java/ti/modules/titanium/gesture/TiDeviceOrientationMonitor.java

```diff
@@ -0,0 +1,631 @@
+	public void setOrientationChangedListener(
+		TiDeviceOrientationMonitor.OrientationChangedListener listener)
+	{
+		this.orientationChangedListener = listener;
+	}
+
+	public TiDeviceOrientationMonitor.OrientationChangedListener getOrientationChangedListener()
+	{
+		return this.orientationChangedListener;
+	}
+
+	public TiDeviceOrientation getLastReadOrientation()
+	{
+		return this.lastReadOrientation;
+	}
+
+	public boolean isRunning()
+	{
+		return (this.isUsingSensorManager || this.isUsingOrientationEventListener);
+	}
+
+	public boolean start()
+	{
+		// Do not continue if already running.
+		if (isRunning()) {
+			return true;
+		}
+
+		// First, attempt to use the accelerometer and gyroscope sensors.
+		this.isUsingSensorManager = startSensor(Sensor.TYPE_ROTATION_VECTOR);
+
+		// If above is not available, then attempt to use accelerometer and magnetic sensors.
+		if (!this.isUsingSensorManager) {
+			boolean wasSensor1Started = startSensor(Sensor.TYPE_ACCELEROMETER);
+			boolean wasSensor2Started = startSensor(Sensor.TYPE_MAGNETIC_FIELD);
+			this.isUsingSensorManager = wasSensor1Started && wasSensor2Started;
+			if (!this.isUsingSensorManager) {
+				if (this.sensorManager != null) {
+					this.sensorManager.unregisterListener(this.sensorEventHandler);
+				}
+			}
+		}
+
+		// If above sensors are not available, then fallback to using Android's OrientationEventListener.
+		// This can only detect device rotation. Does not support face-up or face-down.
+		if (!this.isUsingSensorManager && this.orientationEventHandler.canDetectOrientation()) {
+			this.orientationEventHandler.enable();
+			this.isUsingOrientationEventListener = true;
+		}
+
+		// Enable the screen rotation lock handler.
+		this.screenEventHandler.enable();
+
+		// Return true if we've successfully started an orientation related sensor.
+		return isRunning();
+	}
+
+	public void stop()
+	{
+		// Do not continue if already stopped.
+		if (!isRunning()) {
+			return;
+		}
+
+		// Unregister our "SensorManager" handler if registered.
+		if (this.isUsingSensorManager) {
+			if (this.sensorManager != null) {
+				this.sensorManager.unregisterListener(this.sensorEventHandler);
+			}
+			this.sensorEventHandler.reset();
+			this.isUsingSensorManager = false;
+		}
+
+		// Disable the Android "OrientationEventListener" if running.
+		if (this.isUsingOrientationEventListener) {
+			this.orientationEventHandler.disable();
+			this.isUsingOrientationEventListener = false;
+		}
+
+		// Disable the screen rotation lock handler.
+		this.screenEventHandler.disable();
+	}
+
+	private boolean startSensor(int typeId)
+	{
+		boolean wasSuccessful = false;
+		try {
+			if (this.sensorManager != null) {
+				Sensor sensor = this.sensorManager.getDefaultSensor(typeId);
+				if (sensor != null) {
+					wasSuccessful = this.sensorManager.registerListener(
+							this.sensorEventHandler, sensor,
+							SensorManager.SENSOR_DELAY_NORMAL, this.handler);
+				}
+			}
+		} catch (Exception ex) {
```

Comment:
Should we still log any exceptions here as warnings?

Reply:
I'm already logging a warning in GestureModule here...
https://github.com/appcelerator/titanium_mobile/pull/9585/files#diff-9f99efc31c1562fe3946b86679b014b1R76

The idea is that it'll only log a warning once if it's unable to start the device orientation monitor on startup. If I logged a warning within the DeviceOrientationMonitor class, then we would see a warning logged every time an activty is paused/resumed and I wanted to reduce the noise.